sasl/gssapi/gssapi.go

```diff
@@ -0,0 +1,215 @@
+package gssapi
+
+import (
+	"context"
+	"encoding/asn1"
+	"encoding/binary"
+
+	"github.com/segmentio/kafka-go/sasl"
+
+	"github.com/jcmturner/gokrb5/v8/client"
+	"github.com/jcmturner/gokrb5/v8/crypto"
+	"github.com/jcmturner/gokrb5/v8/gssapi"
+	"github.com/jcmturner/gokrb5/v8/iana/chksumtype"
+	"github.com/jcmturner/gokrb5/v8/iana/keyusage"
+	"github.com/jcmturner/gokrb5/v8/messages"
+	"github.com/jcmturner/gokrb5/v8/types"
+)
+
+const (
+	// https://tools.ietf.org/html/rfc4121#section-4.1
+	TOK_ID_KRB_AP_REQ = "\x01\x00"
+)
+
+type mechanism struct {
+	client      *client.Client
+	serviceName string
+	host        string
+}
+
+func (m mechanism) Name() string {
+	return "GSSAPI"
+}
+
+// Gokrb5v8 uses gokrb5/v8 to implement the GSSAPI mechanism.
+//
+// client is a github.com/gokrb5/v8/client *Client instance.
+// kafkaServiceName is the name of the Kafka service in your Kerberos.
+func Gokrb5v8(client *client.Client, kafkaServiceName string) sasl.Mechanism {
+	return mechanism{client, kafkaServiceName, ""}
+}
+
+// WithHost fulfills the optional sasl.NeedsHost interface
+// needed for this sasl.Mechanism implementation.
+//
+// Unless you are implementing a SASL handshake yourself,
+// you do not need to call this.
+func (m mechanism) WithHost(host string) sasl.Mechanism {
+	m.host = host
```

Comment:
I am curious if these two lines are obvious enough?

The fact that this returns a _copy_ of the GSSAPI mechanism is implicit in using a non-pointer method receiver, and the reason for it is also left implicit.

I think if I were reading this code it would be obvious, so I didn't want to patronizingly draw attention to it, but I am happy to make this more explicit if desired. For example, I can see some benefits to writing it like this instead:

```suggestion
func (m *mechanism) WithHost(host string) sasl.Mechanism {
	// Creating a copy avoids race conditions if the
	// mechanism is reused for multiple connections:
	copy := *m
	copy.host = host
```